### PR TITLE
Add ECDSA support

### LINKED ISSUE #1612: Add ECDSA opcode to txscript
Depends on: https://github.com/kaspanet/go-secp256k1/issues/8
Should use raw SHA256 as the message hash.

### LINKED ISSUE #1621: Add ECDSA addresses
Depends on #1612

### PR DESCRIPTION
**Pull request by someone235**
Closes #1612
Closes #1621

### CODE REVIEW

File: domain/consensus/utils/txscript/opcode.go [hunk 1]
```diff
@@ -2027,6 +2027,89 @@ func opcodeCheckSig(op *parsedOpcode, vm *Engine) error {
...
+	hashType := consensushashing.SigHashType(fullSigBytes[len(fullSigBytes)-1])
+	sigBytes := fullSigBytes[:len(fullSigBytes)-1]
+	if !hashType.IsStandardSigHashType() {
+		return scriptError(ErrInvalidSigHashType, fmt.Sprintf("invalid hash type 0x%x", hashType))
+	}
+	if err := vm.checkSignatureLengthECDSA(sigBytes); err != nil {
+		return err
+	}
+	if err := vm.checkPubKeyEncodingECDSA(pkBytes); err != nil {
+		return err
```
Comment (elichai) [inline domain/consensus/utils/txscript/opcode.go:2070]: I wonder why here we return an error but in the other places we push `false` and return nil

**Review by elichai — APPROVED** [review]
This looks correct from the go-secp usage, haven't fully reviewed though

File: domain/consensus/utils/txscript/opcode.go [hunk 2]
```diff
@@ -2027,6 +2027,89 @@ func opcodeCheckSig(op *parsedOpcode, vm *Engine) error {
...
+	}
+	signature, err := secp256k1.DeserializeECDSASignatureFromSlice(sigBytes)
+	if err != nil {
+		vm.dstack.PushBool(false)
+		return nil
+	}
+
+	var valid bool
+	secpHash := secp256k1.Hash(*sigHash.ByteArray())
+	if vm.sigCache != nil {
```
Comment (stasatdaglabs) [inline domain/consensus/utils/txscript/opcode.go:2093]: This should be `vm.sigCacheECDSA`

File: domain/consensus/utils/txscript/standard.go [hunk 3]
```diff
@@ -208,7 +227,7 @@ func PayToAddrScript(addr util.Address) (*externalapi.ScriptPublicKey, error) {
 		if err != nil {
 			return nil, err
 		}
-		return &externalapi.ScriptPublicKey{script, constants.MaxScriptPublicKeyVersion}, err
+		return &externalapi.ScriptPublicKey{script, 0}, err
```
Comment (stasatdaglabs) [inline domain/consensus/utils/txscript/standard.go:230]: Please extract the 0s in all cases to constants.

File: util/address.go [hunk 4]
```diff
@@ -211,6 +216,68 @@ func (a *AddressPublicKey) String() string {
...
+	publicKey [PublicKeySizeECDSA]byte
+}
+
+// NewAddressPublicKeyECDSA returns a new AddressPublicKeyECDSA. publicKey must be 33
+// bytes.
+func NewAddressPublicKeyECDSA(publicKey []byte, prefix Bech32Prefix) (*AddressPublicKeyECDSA, error) {
+	return newAddressPubKeyECDSA(prefix, publicKey)
+}
+
+// newAddressPubKey is the internal API to create an ECDSA pubkey address
```
Comment (stasatdaglabs) [inline util/address.go:235]: Please update the name of the function in this comment.

File: domain/consensus/processes/transactionvalidator/transactionvalidator_test.go [hunk 5]
```diff
@@ -362,3 +362,127 @@ func TestSigningTwoInputs(t *testing.T) {
 		}
 	})
 }
+
+func TestSigningTwoInputsECDSA(t *testing.T) {
+	testutils.ForAllNets(t, true, func(t *testing.T, params *dagconfig.Params) {
+		params.BlockCoinbaseMaturity = 0
+		factory := consensus.NewFactory()
+		tc, teardown, err := factory.NewTestConsensus(params, false, "TestSigningTwoInputs")
```
Comment (stasatdaglabs) [inline domain/consensus/processes/transactionvalidator/transactionvalidator_test.go:370]: Please update the string.

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

File: domain/consensus/utils/txscript/opcode.go (hunk 1 above)
Comment (someone235) [reply domain/consensus/utils/txscript/opcode.go:2070]: This is copied from OP_CHECKSIG, but I think it's done from malleability reasons

**Review by stasatdaglabs — APPROVED** [review]

**Pull request merged**